lib/api/confighandler.go

```diff
@@ -0,0 +1,336 @@
+// Copyright (C) 2020 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package api
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/crypto/bcrypt"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+type configMuxBuilder struct {
+	*httprouter.Router
+	id  protocol.DeviceID
+	cfg config.Wrapper
+	mut sync.Mutex
+}
+
+func (c *configMuxBuilder) registerConfig(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.RawCopy())
+	})
+	c.PUT(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustConfig(w, r)
+	})
+}
+
+func (c *configMuxBuilder) registerConfigDeprecated(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.RawCopy())
+	})
+	c.POST(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustConfig(w, r)
+	})
+}
+
+func (c *configMuxBuilder) registerConfigInsync(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, map[string]bool{"configInSync": !c.cfg.RequiresRestart()})
+	})
+}
+
+func (c *configMuxBuilder) registerFolders(path string) {
+	c.GET(path, func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.FolderList())
+	})
+	c.PUT(path, func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		var waiter config.Waiter
+		folders := make([]config.FolderConfiguration, 0)
```

Comment:
I don't think you need to `make` it? `var folders []config.FolderConfiguration` should be sufficient to unmarshal into. (Applies below multiple times as well)